### DISCUSSION TITLE
NDI control support

### DISCUSSION THREAD

**Discussion by yatesdr** [glikely/obs-ptz#154] [Ideas]
Love this plugin.    With the move to NDI cameras happening around us, has it been considered to put NDI control on the roadmap?

I understand that NDI licensing is not inherently compatible with GPL, but perhaps with a separate redistributable it could be done.

If it's possible I'm interested in possibly contributing, having just finished another program using NDI control with Python.   I'm pretty rusty at C++ and not up to speed with the OBS plugin architecture though.    The NDI SDK is primarily c, which should mesh well here, and I know that OBS has added NDI viewing capabilities already.

My particular problem with the Visca over TCP was related to buggy camera implementations and closed source, whereas NDI seems to be bulletproof, prompting me to move away from this tool and to a separate control application.

The combination of OBS with NDI control would be quite powerful I think.

**Comment by glikely** [comment]
Hi Yatesdr,

I absolutely would like to add NDI support, and would appreciate some help. This is a personal time project for me, and I don't have a lot of hours to spend on it. Plus I don't actually have any NDI hardware at the moment to work on.

I suspect that adding NDI directly to this plugin isn't the best way to go about it, and not just for the GPL reasons. Instead, I'm thinking thinking about restructuring the PTZ plugin to separate the front end (widgets and configuration) from the back end (protocol implementation), and using the OBS proc handler infrastructure to call between them. This would allow sources to expose a PTZ API that the plugin can call. Then PTZ support could be added directly to the existing NDI plugin.

Anyway, don't let that stop you. I would happily accept patches to add NDI support directly into obs-ptz, and leave the restructuring as a longer term project. Please take a look at the code and let me know if you need help with where to start

**Reply by yatesdr** [reply]
Appreciate that, I will pull it down and have a deeper look.

In NDI the commands are not socket based low-level packets as they are in VISCA-TCP, and the Newtek SDK provides functions in C or CPP that directly send the commands.   The hardest part in python was detecting and getting source references (recv's), after that it was mostly wrappers to call to command the movement and actions.

Assuming this would be used with OBS-NDI plugin, the source list may already have the recv's available, and the library functions may even be included in the runtime already as well.   If this is all true, it could be fairly straight forward to write the integration.

I briefly looked through this project's code, and it looks like a good place to start would be seeing if it's possible to get a recv from the NDI plugin.   Using this it looks like you are pulling video sources to associate with PTZ control based on the selected View.   Can you point me to the general area you're doing that pull?

**Comment by glikely** [comment]
Video sources are associated to a PTZ device by name. There is a
helper function that looks up the associated source using the OBS ABI.
Look here:

https://github.com/glikely/obs-ptz/blob/cc2b689a56b2b9f4fdedca9d162a73661e507320/src/ptz-device.cpp#L638
